### ISSUE TITLE
Race condition with globs and file deletion (ENOENT)

### ISSUE THREAD

**Issue by wchargin** [prettier/prettier#4468]
When passing the unexpanded glob-string `**/*.js` as an argument to
Prettier, a race condition occurs if a file matches the glob when
Prettier expands it, but the file or one of its parent directories has
been deleted by the time that Prettier tries to process it.

The exact error message varies. Usually, I see:

    [error] Invalid configuration file: ENOENT: no such file or directory, scandir '/tmp/tmp.H2DEuUAANZ/bin/commands'

But sometimes, I instead see:

    [error] Unable to read file: bin/commands/test.js
    [error] ENOENT: no such file or directory, open 'bin/commands/test.js'

And, sometimes, I see no error at all, and `prettier` exits
successfully. (Hence: race condition.)

In either case, the exit code is 2, and `bin/commands/test.js` is a file
that no longer exists because `bin/` has been emptied.

I can provide a repro script that works fairly consistently on my
computer—say, triggering a repro more than half the time—but I would not
be surprised if the timing parameters are specific to my machine and
non-portable. That script lives at:
https://gist.github.com/wchargin/3892e168ee7b4ef3bd20ead8b0dc1411

Sample output:
```
Working in: /tmp/tmp.H2DEuUAANZ
Setting up Git...
Setting up Yarn...
Setting up empty.js...
Setting up src...

Setup complete. Version data:
git version 2.16.2
npm version 5.6.0
node version v8.9.4
yarn version 1.5.1
prettier version 1.12.1

Trying to repro...
  * Trying sleep time: 0.17
[error] Invalid configuration file: ENOENT: no such file or directory, scandir '/tmp/tmp.H2DEuUAANZ/bin/commands'
Done! Got exit code: 2
```

**Comment by thorn0** [comment]
@wchargin You didn't write what the expected behavior was. Why even call Prettier on some files you know another process might delete? I'm not sure there is something to fix here.

**Comment by wchargin** [comment]
Expected behavior: Prettier should exit successfully, having prettified
the files that still exist. It may print a clear and non-spammy warning,
like `rsync` does: something like “warning: some files were deleted
before they could be prettified”.

Expected behavior: given the glob `**/*.js` as the sole input argument,
Prettier should never emit an “invalid configuration file” error,
because that configuration *is* in fact valid, and it is Prettier that
is making unsound, racy assumptions about how to interpret it. This
error is especially misleading because the error is in a command-line
argument rather than an actual configuration file.

> Why even call Prettier on some files you know another process might delete?

We’re not calling it with those files directly, and in fact we’ve
instructed Prettier to ignore those files. The real-world case in which
we hit this is a CI pipeline:

  - The pipeline concurrently runs a `lint` job and a `build` job.
  - The `lint` job runs `prettier -l '**/*.js'`, where the Prettier
    config excludes directories like `bin/`.
  - The `build` job builds output into `bin/`, replacing the existing
    contents in the process.

These run concurrently because they *should* be independent: Prettier is
instructed not to look in `bin/` at all, and the build job only writes
to `bin/`. But due to this bug in Prettier, our CI can fail spuriously,
so we had to [work around this with changes to Webpack configs][wa]
(never fun).

We had to keep thinking about this for months until we happened to
refactor our build process in a way that avoided the problem (see
cross-references above).

[wa]: https://github.com/sourcecred/sourcecred/pull/317

**Comment by thorn0** [comment]
Thanks for the explanation. To me, the main point here is this:

> We’re not calling it with those files directly, and in fact we’ve
instructed Prettier to ignore those files

So, it's a real problem with how ignoring files is implemented in CLI. I'm not sure (need to check), but probably if you had excluded those files using a negative glob pattern in the command line instead of using `.prettierignore`, you wouldn't have faced this thing at all.

**Comment by thorn0** [comment]
Fixed long time ago by #6233
